---
output: 
  pdf_document:
    latex_engine: pdflatex
    keep_tex: true
    number_sections: false
params:
  sample_id: "default"
  run_mode: "default"
  prefix: "default"
  quast_output: "default"
  bbmap_output: "default"
  mlst_output: "default"
  rmlst_output: "default"
  kleborate_output: "default"
  amrfinder_output: "default"
  plasmidfinder_output: "default"
  lrefinder_output: "default"
  genome_size: "default"
  kleborate_columns: "default"
  sw_version: "default"
  db_version: "default"
header-includes:
  - \usepackage{booktabs}
  - \usepackage{float}
  - \usepackage{graphicx}
  - \usepackage{enumitem}
  - \usepackage{xcolor}
  - \usepackage{needspace}
  - \usepackage{mathptmx}
  - \setlist[enumerate]{nosep, leftmargin=*, itemsep=0pt, topsep=0pt, partopsep=0pt, parsep=0pt}
geometry: top=1cm, bottom=2cm, left=2cm, right=2cm
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(verbose = TRUE)

library(kableExtra)
library(tidyr)
library(dplyr)
```
```{r set-run-mode, include=FALSE}
# Sequencing platform
run_mode <- params$run_mode
if (run_mode == "kres") {
  sequencing_platform <- "Illumina MiSeq (2x251 bp)"
} else if (run_mode == "eurofins") {
  sequencing_platform <- "Illumina NovaSeq 6000 (2x151 bp)"
} else {
  sequencing_platform <- "Unknown"
}

sample_id <- params$sample_id
```

```{r load-input, include=FALSE}
# Load genome size table
genome_size_tbl <- read.csv($params.genome_size, stringsAsFactors = FALSE)
kleborate_patterns <- read.csv($params.kleborate_columns, stringsAsFactors = FALSE)

# QUAST assembly stats
quast <- read.table($params.quast_output, sep = "\t", fill = TRUE, header = TRUE, comment.char = "")

# BBMap coverage
bbmap <- read.table($params.bbmap_output, sep = "\t", fill = TRUE, header = TRUE, comment.char = "")

# MLST results (original mlst species name)
mlst <- read.table($params.mlst_output, sep = "\t", fill = TRUE, header = TRUE, comment.char = "")

# Ribosomal MLST
rmlst <- read.table($params.rmlst_output, sep = ":", fill = TRUE)

# Kleborate output (if present)
if (file.exists($params.kleborate_output)) {
  kleborate <- read.table($params.kleborate_output, sep = "\t", header = TRUE, stringsAsFactors = FALSE, quote = "", comment.char = "")
  kleborate_patterns <- read.csv($params.kleborate_patterns, header = TRUE)
} else {
  kleborate <- NULL
  kleborate_patterns <- NULL
}

# AMRFinder results
AMR_out <- read.table($params.amrfinder_output, quote = "", comment.char = "", stringsAsFactors = FALSE, header = TRUE, sep = "\t", fill = TRUE)

# LRE-Finder output (for Enterococcus)
if (file.exists($params.lrefinder_output)) {
  lrefinder <- readLines($params.lrefinder_output)
} else {
  lrefinder <- NULL
}

# PlasmidFinder results
plasmidfinder <- read.table($params.plasmidfinder_output, sep = "\t", header = TRUE)

# Software/database versions
if (file.exists($params.sw_versions)) {
  sw_versions <- read.csv($params.sw_versions, header = TRUE)
}
if (file.exists($params.db_version)) {
  db_version <- read.csv($params.db_version, header = TRUE)
}
```

```{r assembly-stats, echo=FALSE, results='asis', warning=FALSE}
quast <- quast %>%
  rename(c("Feature" = "Assembly", "Statistics" = "assembly")) %>%
  filter(Feature == "# contigs (>= 0 bp)" | Feature == "N50" | Feature == "Total length") %>%
  mutate(Feature = gsub("# contigs \\(>= 0 bp\\)", "Number of contigs", Feature))

bbmap <- bbmap %>%
  mutate("Feature" = "coverage") %>%
  rename("Statistics" = "V1") %>%
  select(Feature, Statistics)

assembly_stats <- rbind(quast, bbmap) %>%
  mutate(Statistics = if_else(Feature %in% c("Total length", "Number of contigs", "N50"),
                         as.character(as.integer(Statistics)),
                         as.character(Statistics)))

check_genome_size <- function(species, assembly_size, genome_size_tbl) {
  row <- genome_size_tbl[genome_size_tbl$species == species, ]
  if (nrow(row) == 0) {
    return(NA) # Species not found
  }
  lower <- row$min
  upper <- row$max
  return(assembly_size >= lower & assembly_size <= upper)
}

species <- mlst[,2]
assembly_size <- as.numeric(assembly_stats[assembly_stats$Feature == "Total length", "Statistics"])
within_limits <- check_genome_size(species, assembly_size, genome_size_tbl)

if (is.na(within_limits)) {
  genome_size_comment <- "(Lengde ikke definert i den lokale databasen)"
} else if (within_limits) {
  genome_size_comment <- "(Innenfor  akseptert genomlengde)"
} else {
  row <- genome_size_tbl[genome_size_tbl$species == species, ]
  genome_size_comment <- "(utenfor akseptert genomlengde)"
}

cat("\\begin{tabular}{@{}ll}\n")
cat("Antall kontiger: & ", assembly_stats[1, 2], " \\\\\n")
cat("Assembly-lengde:   & ", assembly_stats[2, 2], " bp ", genome_size_comment, "\\\\\n")
cat("N50 (basert på kontiger \\(\\geq\\) 500 bp):               & ", assembly_stats[3, 2], " bp\\\\\n")
cat("Dekning:          & ", assembly_stats[4, 2], " \\\\\n")
cat("\\end{tabular}\n")
```

```{r mlst, echo=FALSE, results='asis', warning=FALSE}
mlst_species <- mlst[,2]
mlst_st <- mlst[,3]
```

```{r rmlst, echo=FALSE, results='asis', warning=FALSE}
rmlst <- rmlst %>%
  filter(V1 == "Taxon" | V1 == "Support")
```

```{r kleborate, results='asis', echo=FALSE}
if (grepl("Klebsiella", mlst_species) && file.exists($params.kleborate_output)) {
  cat("c) Kleborate (basert på https://github.com/katholt/Kleborate):")
  cat("\\vspace{0.5em}\n")
  kleborate_species <- kleborate[1, "enterobacterales__species__species"]
  kleborate_species_match <- kleborate[1, "enterobacterales__species__species_match"]
  kleborate_QC_warnings <- kleborate[1, "general__contig_stats__QC_warnings"]
  
  cat("\\begin{tabular}{@{}ll}\n")
  cat("Species:                 & \\textbf{", kleborate_species, "} \\\\\n")
  cat("Species match:           & \\textbf{", kleborate_species_match, "} \\\\\n")
  cat("QC warnings:             & \\textbf{", kleborate_QC_warnings, "} \\\\\n")
  cat("\\end{tabular}\n")
}
```

```{r amr-1, include=FALSE}
AMR_out <- AMR_out[, c("Scope", "Type", "Subtype", "Element.symbol", "Class", "Subclass", "Method", "X..Coverage.of.reference", "X..Identity.to.reference")]
colnames(AMR_out) <- c("Scope", "Type", "Subtype", "Element", "Class", "Subclass", "Method", "Coverage", "Identity")
AMR_out$Class <- gsub('/',' ',AMR_out$Class)
AMR_out$Subclass <- gsub("/", " ", AMR_out$Subclass)

AMR_1st_criteria <- AMR_out %>%
  filter(Scope == "core" | Scope == "plus" & Type == "AMR") %>%
  filter(Coverage >= 90 & Identity >= 90) %>%
  arrange(Class, Subclass) %>%
  select(Element, Class, Subclass, Coverage, Identity)

if (nrow(AMR_1st_criteria) == 0) {
    print("Ingen treff som oppfyller første kriterium (90\% dekning og 90\% identitet).")
} else {
    amr-1 <- knitr::kable(
        AMR_1st_criteria,
        format = "latex",
        booktabs = TRUE,
        linesep = "",
        align = "l",
    ) %>% 
    kableExtra::kable_styling(
        latex_options = c("striped", "repeat_header"),
        full_width = FALSE,
    position = "left",
    font_size = 9,
    ) %>%
    column_spec(1, width = "2cm") %>%
    column_spec(2, width = "4cm") %>%
    column_spec(3, width = "5cm") %>%
    column_spec(4, width = "1cm") %>%
    column_spec(5, width = "1cm")
}
```

```{r amr-2, echo=FALSE, results='asis'}
AMR_2nd_criteria <- AMR_out %>%
  filter(Method == "POINTX") %>%
  arrange(Class, Subclass) %>%
  select(Element, Class, Subclass, Coverage, Identity)

if (nrow(AMR_2nd_criteria) == 0) {
    print("Ingen treff som oppfyller andre kriterium (POINTX-metode).")
} else {
    amr-2 <- knitr::kable(
    AMR_2nd_criteria,
        format = "latex",
        booktabs = TRUE,
        linesep = "",
        align = "l",
        ) %>% 
    kableExtra::kable_styling(
        latex_options = c("striped", "repeat_header"),
        full_width = FALSE,
        position = "left",
        font_size = 9,
        ) %>%
    column_spec(1, width = "2cm") %>%
    column_spec(2, width = "4cm") %>%
    column_spec(3, width = "5cm") %>%
    column_spec(4, width = "1cm") %>%
    column_spec(5, width = "1cm")
}
```

```{r amr-3, echo=FALSE, results='asis'}
AMR_3rd_criteria <- AMR_out %>%
  filter(Scope == "core" | Scope == "plus" & Type == "AMR") %>%
  filter(Coverage < 90 | Identity < 90) %>%
  filter(Coverage >= 90 & Identity >= 60 | Coverage >= 60 & Identity >= 90) %>%
  arrange(Class, Subclass) %>%
  select(Element, Class, Subclass, Coverage, Identity)

if (nrow(AMR_3rd_criteria) == 0) {
    print("Ingen treff som oppfyller tredje kriterium (dekning og identitet >60\% < 90\%).")
} else {
    amr-3 <- knitr::kable(
      AMR_3rd_criteria,
        format = "latex",
        booktabs = TRUE,
        linesep = "",
        align = "l",
        ) %>% 
      kableExtra::kable_styling(
        latex_options = c("striped", "repeat_header"),
        full_width = FALSE,
    position = "left",
    font_size = 9,
    ) %>%
    column_spec(1, width = "2cm") %>%
    column_spec(2, width = "4cm") %>%
    column_spec(3, width = "5cm") %>%
    column_spec(4, width = "1cm") %>%
    column_spec(5, width = "1cm")
}
```

```{r amr-4, echo=FALSE, results='asis'}
AMR_4th_criteria <- AMR_out %>%
  filter(Scope == "plus" & !Type == "AMR") %>%
  arrange(Class, Subclass) %>%
  select(Element, Type, Subtype, Class, Subclass, Coverage, Identity)

if (nrow(AMR_4th_criteria) == 0) {
    print("Ingen treff som oppfyller fjerde kriterium (plus scope og ikke AMR type).")
} else {
    amr-4 <- knitr::kable(
    AMR_4th_criteria,
        format = "latex",
        booktabs = TRUE,
        longtable = TRUE,
        linesep = "",
        align = "l",
        ) %>% 
    kableExtra::kable_styling(
        latex_options = c("striped", "repeat_header"),
        full_width = FALSE,
        position = "left",
        font_size = 9,
        ) %>%
    column_spec(1, width = "2cm") %>%
    column_spec(2, width = "2.3cm") %>%
    column_spec(3, width = "2.3cm") %>%
    column_spec(4, width = "2.3cm") %>%
    column_spec(5, width = "2.3cm") %>%
    column_spec(6, width = "1cm") %>%
    column_spec(7, width = "1cm")
}
```

```{r kleborate-tables, results='asis', echo=FALSE}
if (grepl("Klebsiella", mlst_species) && length(kleborate_files) > 0) {
  cat("\\needspace{15\\baselineskip}\n")
  cat("\\textbf{e) Klebsiella-spesifikk resistensanalyse:}\n")
  # Prepare the Kleborate data
  kleborate_long <- tibble(
    Pattern = names(kleborate),
    Value = as.character(unlist(kleborate[1, ]))
  )
  
  kleborate_long <- kleborate_long %>%
    mutate(Pattern = gsub(".*__", "", Pattern))
  
  kleborate_long_filtered <- kleborate_long[kleborate_long$Pattern %in% kleborate_patterns$Pattern, ]
  kleborate_long_filtered_metadata <- kleborate_long_filtered %>%
    left_join(kleborate_patterns, by = "Pattern") %>%
    arrange(Type)
  
  # Separate tables by Type
  mutations_table <- kleborate_long_filtered_metadata %>%
    filter(Type == "mutations") %>%
    select(Pattern, Value)
  
  virulence_table <- kleborate_long_filtered_metadata %>%
    filter(Type == "virulence") %>%
    select(Pattern, Value)
  
  locus_table <- kleborate_long_filtered_metadata %>%
    filter(Type %in% c("K_locus", "O_locus")) %>%
    select(Pattern, Value)
  
  # Render the mutations table
  cat("\\vspace{1em}\n\\par\n")
  cat("\\needspace{15\\baselineskip}\n")
  cat("\\textbf{Mutasjoner:}\n")
  knitr::kable(
    mutations_table,
    format = "latex",
    booktabs = TRUE,
    longtable = FALSE,
    linesep = "",
    align = "l"
  ) %>%
    kableExtra::kable_styling(
      latex_options = c("striped", "repeat_header"),
      full_width = FALSE,
      position = "left",
      font_size = 9
    ) %>%
    column_spec(1, width = "5cm") %>%
    column_spec(2, width = "4cm") %>%
    print()
    cat("Kolistinresistens er bestemt av trunkering av/eller manglende mgrB- og pmrB-kjernegener. OmpK35- og OmpK36-resistens er bestemt av trunkering eller punktmutasjoner. Trunkering representeres av % lengden av aminosyrer fra kodon start.\n\\par\n")
  
  # Render the virulence table
  cat("\\vspace{1em}\n")
  cat("\\needspace{15\\baselineskip}\n")
  cat("\\textbf{Virulensfaktorer:}\n")
  knitr::kable(
    virulence_table,
    format = "latex",
    booktabs = TRUE,
    longtable = FALSE,
    linesep = "",
    align = "l"
  ) %>%
    kableExtra::kable_styling(
      latex_options = c("striped", "repeat_header"),
      full_width = FALSE,
      position = "left",
      font_size = 9
    ) %>%
    column_spec(1, width = "5cm") %>%
    column_spec(2, width = "4cm") %>%
    print()
    cat("0 = negative for all of yersiniabactin (ybt), colibactin (clb), aerobactin (iuc); 1 = yersiniabactin only; 2 = yersiniabactin and colibactin (or colibactin only); 3 = aerobactin (without yersiniabactin or colibactin); 4 = aerobactin with yersiniabactin (without colibactin); 5 = yersiniabactin, colibactin and aerobactin \n\\par\n")
  
  # Render the O and K locus table
  cat("\\vspace{1em}\n")
  cat("\\needspace{15\\baselineskip}\n")
  cat("\\textbf{O- og K-locus:}\n")
  knitr::kable(
    locus_table,
    format = "latex",
    booktabs = TRUE,
    longtable = FALSE,
    linesep = "",
    align = "l"
  ) %>%
    kableExtra::kable_styling(
      latex_options = c("striped", "repeat_header"),
      full_width = FALSE,
      position = "left",
      font_size = 9
    ) %>%
    column_spec(1, width = "5cm") %>%
    column_spec(2, width = "4cm") %>%
    print()
}
```

```{r lre-finder-tables, results='asis', echo=FALSE}
if (grepl("Enterococcus", mlst_species)) {
  cat("\\needspace{20\\baselineskip}\n")
  cat("\\textbf{e) Linezolid Resistens: (based på LRE-finder, https://cge.cbs.dtu.dk/services/LRE-finder/)}")
  cat("\\vspace{1em}\n")

  # Species identified
  lrefinder_species <- which(grepl("Species identified:", lrefinder)) + 1
  lrefinder_species <- lrefinder[lrefinder_species]

  cat("\\begin{tabular}{@{}ll}\n")
  cat("Species:                 & \\textbf{", lrefinder_species, "} \\\\\n")
  cat("\\end{tabular}\n")

  # Find section starts
  genes_start <- which(grepl("Genes Identified:", lrefinder)) + 1
  mut_start <- which(grepl("Identified mutations in 23s:", lrefinder)) + 1

  cat("\\vspace{1em}\n")
  cat("Gener:\n")

  # Extract gene hits
  gene_lines <- lrefinder[genes_start:(mut_start - 3)]
  gene_lines <- gene_lines[gene_lines != ""]
  if (length(gene_lines) > 1) {  # header + at least one data line
    gene_df <- read.table(text = paste(gene_lines, collapse = "\n"), header = TRUE, stringsAsFactors = FALSE)
    lre_gene <- knitr::kable(
      gene_df,
      format = "latex",
      booktabs = TRUE,
      longtable = FALSE,
      linesep = "",
      align = "c"
    ) %>%
      kableExtra::kable_styling(
        latex_options = c("striped", "repeat_header"),
        full_width = FALSE,
        position = "left",
        font_size = 9
      ) %>%
      column_spec(1, width = "4cm") %>%
      column_spec(2, width = "2cm") %>%
      column_spec(3, width = "2cm")
    print(lre_gene)

  } else {
    cat("Ingen treff\n")
  }

  # Extract mutations
  cat("Mutations:\n")
  mut_lines <- lrefinder[(mut_start + 1):length(lrefinder)]
  mut_lines <- mut_lines[mut_lines != "" & !grepl("reference", mut_lines)]
  if (length(mut_lines) > 0) {
    mut_df <- read.table(text = paste(mut_lines, collapse = "\n"), header = FALSE, stringsAsFactors = FALSE)
    colnames(mut_df) <- c("Position", "WildTypeRatio", "MutantRatio", "Phenotype")
    lre_mut <- knitr::kable(
      mut_df,
      format = "latex",
      booktabs = TRUE,
      longtable = FALSE,
      linesep = "",
      align = "c"
    ) %>%
      kableExtra::kable_styling(
        latex_options = c("striped", "repeat_header"),
        full_width = FALSE,
        position = "left",
        font_size = 9
      ) %>%
      column_spec(1, width = "2cm") %>%
      column_spec(2, width = "2cm") %>%
      column_spec(3, width = "1.5cm") %>%
      column_spec(4, width = "1.5cm")
    print(lre_mut)
  } else {
    print("Ingen treff")
  }
}
```

```{r results='asis', echo=FALSE}
plasmidfinder <- plasmidfinder %>%
  select(Plasmid, Identity)

if (length(plasmidfinder) > 0) {
  plasmid_tbl <- knitr::kable(
    plasmidfinder,
    format = "latex",
    booktabs = TRUE,
    longtable = FALSE,
    linesep = "",
    align = "c"
  ) %>%
    kableExtra::kable_styling(
      latex_options = c("striped", "repeat_header"),
      full_width = FALSE,
    position = "left",
    font_size = 9
  ) %>%
  column_spec(1, width = "3cm") %>%
  column_spec(2, width = "3cm")
print(plasmid_tbl)
} else {
  cat("Ingen plasmidreplikon funnet.\n")
}
```

```{r results='asis', echo=FALSE}
sw_tbl <- knitr::kable(
  sw_versions,
  format = "latex",
  booktabs = TRUE,
  longtable = FALSE,
  linesep = "",
  align = "c"
) %>%
  kableExtra::kable_styling(
    latex_options = c("striped", "repeat_header"),
    full_width = FALSE,
    position = "left",
    font_size = 9
  ) %>%
  column_spec(1, width = "3cm") %>%
  column_spec(2, width = "3cm")
print(sw_tbl)

db_tbl <- knitr::kable(
  db_version,
  format = "latex",
  booktabs = TRUE,
  longtable = FALSE,
  linesep = "",
  align = "c"
) %>%
  kableExtra::kable_styling(
    latex_options = c("striped", "repeat_header"),
    full_width = FALSE,
    position = "left",
    font_size = 9
  ) %>%
  column_spec(1, width = "3cm") %>%
  column_spec(2, width = "3cm")
print(db_tbl)
```



\begin{figure}[H]
\includegraphics[width=0.6\textwidth]{/bigdata/Jessin/Scripts/illumina_kres/assets/NS_paavisning_anitibiotikaresistens_RGB_cropped.png}
\centering
\end{figure}

\noindent\textcolor{violet}{\rule{\textwidth}{0.7pt}}
\subsection*{\centering \color{violet}Bioinformatisk rapport for helgenomssekvensering}
\noindent\textcolor{violet}{\rule{\textwidth}{0.7pt}}

\vspace{1em}
`r format(Sys.Date(), "%d.%m.%Y")`

\vspace{1em}
\textbf{\color{violet}RID : `r sample_id`} 

\subsection*{\color{violet}Metodesammendrag:}

Genomisk DNA sekvensert med `r sequencing_platform`. Assembly utført via Shovill (med SPAdes assembler), maks dekning satt til 150x og kontiger < 200 bp og 2x dekning fjernet. Species, MLST-type, resistensgener og plasmidreplikoner identifisert ved bruk av intern bioinformatisk pipeline.

Følgende kvalitetskontrollkriterier gjelder:

\begin{enumerate}[nosep, leftmargin=*, itemsep=0pt, topsep=0pt, partopsep=0pt, parsep=0pt]
  \item \textbf{Dekning:} $\,\geq\,$ \textbf{40x}
  \item \textbf{Genomlengde:} $\,\geq\,$ \textbf{95\%} og $\,\leq\,$ \textbf{105\%} av minste/største lukkede genom av samme species i NCBI GenBank
  \item \textbf{Maks antall kontiger:} \textbf{400}
\end{enumerate}
\noindent\rule{\textwidth}{0.5pt}

\subsection*{\color{violet}Resultater:}

\subsubsection*{\color{blue}1. Assembly statistikk:}
\vspace{1em}
```{r assembly-stats, echo=FALSE, results='asis'}
cat("\\begin{tabular}{@{}ll}\n")
cat("Antall kontiger: & ", assembly_stats[1, 2], " \\\\\n")
cat("Assembly-lengde:   & ", assembly_stats[2, 2], " bp ", genome_size_comment, "\\\\\n")
cat("N50 (basert på kontiger \\(\\geq\\) 500 bp):               & ", assembly_stats[3, 2], " bp\\\\\n")
cat("Dekning:          & ", assembly_stats[4, 2], " \\\\\n")
cat("\\end{tabular}\n")
```

\subsubsection*{\color{blue}2. Taksonomisk klassifisering:}
\vspace{1em}

a) Species og MLST-type (basert på https://pubmlst.org/):
**`r mlst_species` ST-`r mlst_st`**
\vspace{1em}

b) Ribosomal MLST - Species og støtteprosent (basert på https://pubmlst.org/species-id):
**`r rmlst[1,2]`**  **`r paste0("(", rmlst[2,2], ")")`**
\vspace{1em}

```{r print-kleborate-species, echo=FALSE, results='asis'}
if (grepl("Klebsiella", mlst_species) && file.exists($params.kleborate_output)) {
  cat("c) Kleborate (basert på https://github.com/katholt/Kleborate):")
  cat("\\vspace{0.5em}\n")
  cat("\\begin{tabular}{@{}ll}\n")
  cat("Species:                 & \\textbf{", kleborate_species, "} \\\\\n")
  cat("Species match:           & \\textbf{", kleborate_species_match, "} \\\\\n")
  cat("QC warnings:             & \\textbf{", kleborate_QC_warnings, "} \\\\\n")
  cat("\\end{tabular}\n")
}
```

\newpage

\subsubsection*{\color{blue}3. Resistens:}
\vspace{1em}
Resistensgener basert på kjernedatabasen av Bacterial Antimicrobial Resistance Reference Gene Database BioProject PRJNA313047. Inkluderer også gener som gir resistens mot ikke-antibiotika (f.eks. kvartære ammoniumforbindelser) og antibiotika som brukes i veterinærmedisin.
\vspace{1em}
\textbf{a) Analysekriterier:} $\,\geq\,$ 90\% dekning og $\,\geq\,$ 90\% nukleotididentitet  
 
