---
output:
  pdf_document:
    latex_engine: pdflatex
    keep_tex: true
    number_sections: false
params:
  sample_id: ""
  run_mode: ""
  genome_size: ""
  kleborate_columns: ""
  quast_out: ""
  bbmap_out: ""
  mlst_out: ""
  rmlst_out: ""
  kleborate_out: ""
  amrfinder_out: ""
  plasmidfinder_out: ""
  lrefinder_out: ""
  versions: ""
  kres_logo: ""
header-includes:
  - \usepackage{booktabs}
  - \usepackage{float}
  - \usepackage{graphicx}
  - \usepackage{enumitem}
  - \usepackage{xcolor}
  - \usepackage{needspace}
  - \usepackage{mathptmx}
  - \setlist[enumerate]{nosep, leftmargin=*, itemsep=0pt, topsep=0pt, partopsep=0pt, parsep=0pt}
geometry: top=1cm, bottom=2cm, left=2cm, right=2cm
---
\begin{figure}[H]
\includegraphics[width=0.6\textwidth]{`r params$kres_logo`}
\centering
\end{figure}

\noindent\textcolor{violet}{\rule{\textwidth}{0.7pt}}
\subsection*{\centering \color{violet}Bioinformatisk rapport for helgenomssekvensering}
\noindent\textcolor{violet}{\rule{\textwidth}{0.7pt}}

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(verbose = TRUE)

library(kableExtra)
library(tidyr)
library(dplyr)
library(tibble)

# Sequencing platform
run_mode <- params$run_mode
if (run_mode == "kres") {
  sequencing_platform <- "Illumina MiSeq (2x250 bp)"
} else if (run_mode == "eurofins") {
  sequencing_platform <- "Illumina NovaSeq 6000 (2x150 bp)"
} else if (run_mode == "ous") {
  sequencing_platform <- "Illumina MiSeq (2x300 bp)"
} else {
  sequencing_platform <- "Unknown"
}

sample_id <- params$sample_id
```

```{r load-input-files, include=FALSE}
# Load genome size table
genome_size_tbl <- read.csv(params$genome_size, stringsAsFactors = FALSE)
kleborate_patterns <- read.csv(params$kleborate_columns, stringsAsFactors = FALSE)

# QUAST assembly stats
quast <- read.table(params$quast_out, sep = "\t", fill = TRUE, header = TRUE, comment.char = "")

# BBMap coverage
bbmap <- readLines(file.path(params$bbmap_out))


# MLST results (original mlst species name)
mlst <- read.table(params$mlst_out, sep = "\t")


# Ribosomal MLST
rmlst <- read.table(params$rmlst_out, sep = ":", fill = TRUE)

# Kleborate
kleborate <- NULL
if (file.exists(params$kleborate_out) &&
    file.size(params$kleborate_out) > 0) {

    kleborate <- read.table(file.path(params$kleborate_out), sep = "\t", header = TRUE, stringsAsFactors = FALSE, quote = "", comment.char = "")
}

# AMRFinder results
AMR_out <- read.table(params$amrfinder_out, quote = "", comment.char = "", stringsAsFactors = FALSE, header = TRUE, sep = "\t", fill = TRUE)

# Kleborate patterns (if used)
if (file.exists(params$kleborate_columns)) {
  kleborate_patterns <- read.csv(params$kleborate_columns, header = TRUE)
}

# LRE-Finder output (for Enterococcus)
lrefinder <- NULL
if (file.exists(params$lrefinder_out) &&
    file.size(params$lrefinder_out) > 0) {

    lrefinder <- readLines(file.path(params$lrefinder_out))
}


# PlasmidFinder results
plasmidfinder <- read.table(file.path(params$plasmidfinder_out), sep = "\t", header = TRUE)

# Software versions
yml <- yaml::read_yaml(params$versions)  # replace with your filename

# Optionally skip the Workflow section
yml <- yml[names(yml) != "Workflow"]


sw_df <- enframe(unlist(yml)) %>%
  separate(name, into = c("module", "Name"), fill = "right", sep = "\\.") %>%
  rename(Version = value) %>%
  select(-module) %>%
  filter(!Name == "samtools") %>%
  filter(!Name == "pigz") %>%
  filter(!Name == "amrfinderplus-database") %>%
  filter(!Name == "PlasmidFinder database")

db_df <- enframe(unlist(yml)) %>%
  separate(name, into = c("module", "Name"), fill = "right", sep = "\\.") %>%
  rename(Version = value) %>%
  select(-module) %>%
  filter(Name == "amrfinderplus-database" | Name == "PlasmidFinder database")
```
\vspace{1em}
\textbf{\textcolor{violet}{RID : `r sample_id`}}

\subsection*{\color{violet}Metodesammendrag:}

Genomisk DNA sekvensert med `r sequencing_platform`. Assembly utført via Shovill (med SPAdes assembler), maks dekning satt til 150x og kontiger < 200 bp og 2x dekning fjernet. Species, MLST-type, resistensgener og plasmidreplikoner identifisert ved bruk av intern bioinformatisk pipeline.

Følgende kvalitetskontrollkriterier gjelder:

\begin{enumerate}[nosep, leftmargin=*, itemsep=0pt, topsep=0pt, partopsep=0pt, parsep=0pt]
  \item \textbf{Dekning:} $\,\geq\,$ \textbf{40x}
  \item \textbf{Genomlengde:} $\,\geq\,$ \textbf{95\%} og $\,\leq\,$ \textbf{105\%} av minste/største lukk>
  \item \textbf{Maks antall kontiger:} \textbf{400}
\end{enumerate}
\noindent\rule{\textwidth}{0.5pt}

\subsection*{\color{violet}Resultater:}

\subsubsection*{\color{blue}1. Assembly statistikk:}
\vspace{1em}
```{r assembly-stats, echo=FALSE, results='asis', warning=FALSE}
quast <- quast %>%
  rename(c("Feature" = "Assembly", "Statistics" = "contigs")) %>%
  filter(Feature == "# contigs (>= 0 bp)" | Feature == "N50" | Feature == "Total length") %>%
  mutate(Feature = gsub("# contigs \\(>= 0 bp\\)", "Number of contigs", Feature))

bbmap_cov <- which(grepl("Average coverage:", bbmap))
bbmap_cov <- data.frame(
  Feature = "Coverage",
  Statistics = gsub("Average coverage: ", "", bbmap[bbmap_cov])
)


assembly_stats <- rbind(quast, bbmap_cov) %>%
  mutate(Statistics = if_else(Feature %in% c("Total length", "Number of contigs", "N50"),
                         as.character(as.integer(Statistics)),
                         as.character(Statistics)))

check_genome_size <- function(species, assembly_size, genome_size_tbl) {
  row <- genome_size_tbl[genome_size_tbl$species == species, ]
  if (nrow(row) == 0) {
    return(NA) # Species not found
  }
  lower <- row$min
  upper <- row$max
  return(assembly_size >= lower & assembly_size <= upper)
}

species <- mlst[,2]
assembly_size <- as.numeric(assembly_stats[assembly_stats$Feature == "Total length", "Statistics"])
within_limits <- check_genome_size(species, assembly_size, genome_size_tbl)

if (is.na(within_limits)) {
  genome_size_comment <- "(Lengde ikke definert i den lokale databasen)"
} else if (within_limits) {
  genome_size_comment <- "(Innenfor  akseptert genomlengde)"
} else {
  row <- genome_size_tbl[genome_size_tbl$species == species, ]
  genome_size_comment <- "(utenfor akseptert genomlengde)"
}

cat("\\begin{tabular}{@{}ll}\n")
cat("Antall kontiger: & ", assembly_stats[1, 2], " \\\\\n")
cat("Assembly-lengde:   & ", assembly_stats[2, 2], " bp ", genome_size_comment, "\\\\\n")
cat("N50 (basert på kontiger \\(\\geq\\) 500 bp):               & ", assembly_stats[3, 2], " bp\\\\\n")
cat("Dekning:          & ", assembly_stats[4, 2], " \\\\\n")
cat("\\end{tabular}\n")
```
\subsubsection*{\color{blue}2. Taksonomisk klassifisering:}
\vspace{1em}

a) Species og MLST-type (basert på https://pubmlst.org/):
```{r mlst, echo=FALSE, results='asis', warning=FALSE}
mlst_species <- mlst[,2]
mlst_st <- mlst[,3]
```
**`r mlst_species` ST-`r mlst_st`**
\vspace{1em}

b) Ribosomal MLST - Species og støtteprosent (basert på https://pubmlst.org/species-id):
```{r rmlst, echo=FALSE, results='asis', warning=FALSE}
rmlst <- rmlst %>%
  filter(V1 == "Taxon" | V1 == "Support")
```
**`r rmlst[1,2]`**  **`r paste0("(", rmlst[2,2], ")")`**
\vspace{1em}

```{r results='asis', echo=FALSE}
if (!is.null(kleborate) && nrow(kleborate) == 1 ) {
  cat("c) Kleborate (basert på https://github.com/katholt/Kleborate):")
  cat("\\vspace{0.5em}\n")
  kleborate_species <- kleborate[1, "enterobacterales__species__species"]
  kleborate_species_match <- kleborate[1, "enterobacterales__species__species_match"]
  kleborate_QC_warnings <- kleborate[1, "general__contig_stats__QC_warnings"]

  cat("\\begin{tabular}{@{}ll}\n")
  cat("Species:                 & \\textbf{", kleborate_species, "} \\\\\n")
  cat("Species match:           & \\textbf{", kleborate_species_match, "} \\\\\n")
  cat("QC warnings:             & \\textbf{", kleborate_QC_warnings, "} \\\\\n")
  cat("\\end{tabular}\n")
}
```
\newpage

\subsubsection*{\color{blue}3. Resistens:}
\vspace{0.5em}
Resistensgener basert på kjernedatabasen av Bacterial Antimicrobial Resistance Reference Gene Database BioProject PRJNA313047. Inkluderer også gener som gir resistens mot ikke-antibiotika (f.eks. kvartære ammoniumforbindelser) og antibiotika som brukes i veterinærmedisin.

```{r, include=FALSE}
AMR_out <- AMR_out[, c("Scope", "Type", "Subtype", "Element.symbol", "Class", "Subclass", "Method", "X..Coverage.of.reference", "X..Identity.to.reference")]
colnames(AMR_out) <- c("Scope", "Type", "Subtype", "Element", "Class", "Subclass", "Method", "Coverage", "Identity")
AMR_out$Class <- gsub('/',' ',AMR_out$Class)
AMR_out$Subclass <- gsub("/", " ", AMR_out$Subclass)

AMR_1st_criteria <- AMR_out %>%
  filter(Scope == "core" | Scope == "plus" & Type == "AMR") %>%
  filter(Method != "POINTX") %>%
  filter(Coverage >= 90 & Identity >= 90) %>%
  arrange(Class, Subclass) %>%
  select(Element, Class, Subclass, Coverage, Identity)
```
\textbf{a) Analysekriterier:} $\,\geq\,$ 90\% dekning og $\,\geq\,$ 90\% nukleotididentitet

```{r AMRFinder_test, echo=FALSE, results='asis'}
knitr::kable(
    AMR_1st_criteria,
    format = "latex",
    booktabs = TRUE,
    linesep = "",
    align = "l",
    ) %>%
  kableExtra::kable_styling(
    latex_options = c("striped", "repeat_header"),
    full_width = FALSE,
    position = "left",
    font_size = 9,
    ) %>%
  column_spec(1, width = "2cm") %>%
  column_spec(2, width = "4cm") %>%
  column_spec(3, width = "5cm") %>%
  column_spec(4, width = "1cm") %>%
  column_spec(5, width = "1cm")
```

\vspace{1em}
\needspace{15\baselineskip}
\textbf{b) Mutasjonsmediert resistens:}

```{r, echo=FALSE, results='asis'}
AMR_2nd_criteria <- AMR_out %>%
  filter(Method == "POINTX") %>%
  arrange(Class, Subclass) %>%
  select(Element, Class, Subclass, Coverage, Identity)

if (nrow(AMR_2nd_criteria) > 0) {
    knitr::kable(
    AMR_2nd_criteria,
        format = "latex",
        booktabs = TRUE,
        linesep = "",
        align = "l",
        ) %>%
    kableExtra::kable_styling(
        latex_options = c("striped", "repeat_header"),
        full_width = FALSE,
        position = "left",
        font_size = 9,
        ) %>%
    column_spec(1, width = "2cm") %>%
    column_spec(2, width = "4cm") %>%
    column_spec(3, width = "5cm") %>%
    column_spec(4, width = "1cm") %>%
    column_spec(5, width = "1cm")
} else {
    cat("\\vspace{0.5em}")
    cat("\\textbf{Ingen treff}\n")
    cat("\\vspace{0.5em}")
}
```
\vspace{1em}
\needspace{15\baselineskip}
\textbf{c) Analysekriterier:} $\,\geq\,$ 60\% - $\,\leq\,$ 90\% dekning og $\,\geq\,$ 60\% - $\,\leq\,$ 90>

```{r, echo=FALSE, results='asis'}
AMR_3rd_criteria <- AMR_out %>%
  filter(Scope == "core" | Scope == "plus" & Type == "AMR") %>%
  filter(Coverage < 90 | Identity < 90) %>%
  filter(Coverage >= 90 & Identity >= 60 | Coverage >= 60 & Identity >= 90) %>%
  arrange(Class, Subclass) %>%
  select(Element, Class, Subclass, Coverage, Identity)

knitr::kable(
  AMR_3rd_criteria,
    format = "latex",
    booktabs = TRUE,
    linesep = "",
    align = "l",
    ) %>%
  kableExtra::kable_styling(
    latex_options = c("striped", "repeat_header"),
    full_width = FALSE,
    position = "left",
    font_size = 9,
    ) %>%
  column_spec(1, width = "2cm") %>%
  column_spec(2, width = "4cm") %>%
  column_spec(3, width = "5cm") %>%
  column_spec(4, width = "1cm") %>%
  column_spec(5, width = "1cm")
```
\vspace{1em}
\needspace{15\baselineskip}
\textbf{d) Gener koblet til resistens, biocider, stress, effluks, virulens:} $\,\geq\,$ 60\% dekning og og $\,\geq\,$ 60\% nukleotididentitet.
Gener basert på Plus databasen av Bacterial Antimicrobial Resistance Reference Gene Database BioProject PRJNA313047.

```{r, echo=FALSE, results='asis'}
AMR_4th_criteria <- AMR_out %>%
  filter(Scope == "plus" & !Type == "AMR") %>%
  arrange(Class, Subclass) %>%
  select(Element, Type, Subtype, Class, Subclass, Coverage, Identity)

knitr::kable(
  AMR_4th_criteria,
    format = "latex",
    booktabs = TRUE,
    longtable = TRUE,
    linesep = "",
    align = "l",
    ) %>%
  kableExtra::kable_styling(
    latex_options = c("striped", "repeat_header"),
    full_width = FALSE,
    position = "left",
    font_size = 9,
    ) %>%
  column_spec(1, width = "2cm") %>%
  column_spec(2, width = "2.3cm") %>%
  column_spec(3, width = "2.3cm") %>%
  column_spec(4, width = "2.3cm") %>%
  column_spec(5, width = "2.3cm") %>%
  column_spec(6, width = "1cm") %>%
  column_spec(7, width = "1cm")
```

\vspace{1em}
```{r results='asis', echo=FALSE}
if (!is.null(kleborate) && nrow(kleborate) == 1) {
    cat("\\needspace{15\\baselineskip}\n")
    cat("\\textbf{e) Klebsiella-spesifikk resistensanalyse:}\n")
    # Prepare the Kleborate data
    kleborate_long <- tibble(
        Pattern = names(kleborate),
        Value = as.character(unlist(kleborate[1, ]))
    )
    kleborate_long <- kleborate_long %>%
        mutate(Pattern = gsub(".*__", "", Pattern))
    kleborate_long_filtered <- kleborate_long[kleborate_long$Pattern %in% kleborate_patterns$Pattern, ]
    kleborate_long_filtered_metadata <- kleborate_long_filtered %>%
        left_join(kleborate_patterns, by = "Pattern") %>%
        arrange(Type)

    # Separate tables by Type
    mutations_table <- kleborate_long_filtered_metadata %>%
        filter(Type == "mutations") %>%
        select(Pattern, Value)

    virulence_table <- kleborate_long_filtered_metadata %>%
        filter(Type == "virulence") %>%
        select(Pattern, Value)

    locus_table <- kleborate_long_filtered_metadata %>%
        filter(Type %in% c("K_locus", "O_locus")) %>%
        select(Pattern, Value)

    # Render the mutations table
    cat("\\vspace{1em}\n\\par\n")
    cat("\\needspace{10\\baselineskip}\n")
    cat("\\textbf{Mutasjoner:}\n")
    knitr::kable(
        mutations_table,
        format = "latex",
        booktabs = TRUE,
        longtable = FALSE,
        linesep = "",
        align = "l"
    ) %>%
        kableExtra::kable_styling(
            latex_options = c("striped", "repeat_header"),
            full_width = FALSE,
            position = "left",
            font_size = 9
            ) %>%
            column_spec(1, width = "5cm") %>%
            column_spec(2, width = "4cm") %>%
        print()
        cat("Kolistinresistens er bestemt av trunkering av/eller manglende mgrB- og pmrB-kjernegener. OmpK35- og OmpK36-resistens er bestemt av trunkering eller punktmutasjoner. Trunkering representeres av % lengden av aminosyrer fra kodon start.\n\\par\n")

    # Render the virulence table
    cat("\\vspace{1em}\n")
    cat("\\needspace{15\\baselineskip}\n")
    cat("\\textbf{Virulensfaktorer:}\n")
    knitr::kable(
        virulence_table,
        format = "latex",
        booktabs = TRUE,
        longtable = FALSE,
        linesep = "",
        align = "l"
    ) %>%
        kableExtra::kable_styling(
            latex_options = c("striped", "repeat_header"),
            full_width = FALSE,
            position = "left",
            font_size = 9
            ) %>%
            column_spec(1, width = "5cm") %>%
            column_spec(2, width = "4cm") %>%
        print()
        cat("0 = negative for all of yersiniabactin (ybt), colibactin (clb), aerobactin (iuc); 1 = yersiniabactin only; 2 = yersiniabactin and colibactin (or colibactin only); 3 = aerobactin (without yersiniabactin or colibactin); 4 = aerobactin with yersiniabactin (without colibactin); 5 = yersiniabactin, colibactin and aerobactin \n\\par\n")

    # Render the O and K locus table
    cat("\\vspace{1em}\n")
    cat("\\needspace{15\\baselineskip}\n")
    cat("\\textbf{O- og K-locus:}\n")
    knitr::kable(
        locus_table,
        format = "latex",
        booktabs = TRUE,
        longtable = FALSE,
        linesep = "",
        align = "l"
    ) %>%
        kableExtra::kable_styling(
            latex_options = c("striped", "repeat_header"),
            full_width = FALSE,
            position = "left",
            font_size = 9
        ) %>%
        column_spec(1, width = "5cm") %>%
        column_spec(2, width = "4cm") %>%
        print()
}
```

\vspace{1em}
```{r results='asis', echo=FALSE}
if (!is.null(lrefinder)) {
  cat("\\needspace{20\\baselineskip}\n")
  cat("\\textbf{e) Linezolid Resistance: (påvist med LRE-finder, https://cge.cbs.dtu.dk/services/LRE-finder/)}")
  cat("\\vspace{1em}\n")

  # Species identified
  lrefinder_species <- which(grepl("Species identified:", lrefinder)) + 1
  lrefinder_species <- lrefinder[lrefinder_species]

  cat("\\begin{tabular}{@{}ll}\n")
  cat("Species:                 & \\textbf{", lrefinder_species, "} \\\\\n")
  cat("\\end{tabular}\n")

  # Find section starts
  genes_start <- which(grepl("Genes Identified:", lrefinder)) + 1
  mut_start <- which(grepl("Identified mutations in 23s:", lrefinder)) + 1

  cat("\\vspace{1em}\n")
  cat("Gener:\n")

  # Extract gene hits
  gene_lines <- lrefinder[genes_start:(mut_start - 3)]
  gene_lines <- gene_lines[gene_lines != ""]
  if (length(gene_lines) > 1) {  # header + at least one data line
    gene_df <- read.table(text = paste(gene_lines, collapse = "\n"), header = TRUE, stringsAsFactors = FALSE)
    lre_gene <- knitr::kable(
      gene_df,
      format = "latex",
      booktabs = TRUE,
      longtable = FALSE,
      linesep = "",
      align = "c"
    ) %>%
      kableExtra::kable_styling(
        latex_options = c("striped", "repeat_header"),
        full_width = FALSE,
        position = "left",
        font_size = 9
      ) %>%
      column_spec(1, width = "4cm") %>%
      column_spec(2, width = "2cm") %>%
      column_spec(3, width = "2cm")
    print(lre_gene)

  } else {
    cat("\\textbf{Ingen treff}\n")
  }

  # Extract mutations
  cat("Mutations:\n")
  mut_lines <- lrefinder[(mut_start + 1):length(lrefinder)]
  mut_lines <- mut_lines[mut_lines != "" & !grepl("reference", mut_lines)]
  if (length(mut_lines) > 0) {
    mut_df <- read.table(text = paste(mut_lines, collapse = "\n"), header = FALSE, stringsAsFactors = FALSE)
    colnames(mut_df) <- c("Position", "WildTypeRatio", "MutantRatio", "Phenotype")
    lre_mut <- knitr::kable(
      mut_df,
      format = "latex",
      booktabs = TRUE,
      longtable = FALSE,
      linesep = "",
      align = "c"
    ) %>%
      kableExtra::kable_styling(
        latex_options = c("striped", "repeat_header"),
        full_width = FALSE,
        position = "left",
        font_size = 9
      ) %>%
      column_spec(1, width = "2cm") %>%
      column_spec(2, width = "2cm") %>%
      column_spec(3, width = "1.5cm") %>%
      column_spec(4, width = "1.5cm")
    print(lre_mut)
  } else {
    cat("\\textbf{Ingen treff}\n")
  }
}
```

\needspace{10\baselineskip}
\subsubsection*{\color{blue}4. Plasmidreplikon:}
\vspace{0.5em}
(påvist med PlasmidFinder https://cge.cbs.dtu.dk/services/PlasmidFinder/). Analysen gjelder bare Enterobacterales og Gram-positive.

```{r results='asis', echo=FALSE}
plasmidfinder <- plasmidfinder %>%
  select(Plasmid, Identity)

plasmid_tbl <- knitr::kable(
  plasmidfinder,
  format = "latex",
  booktabs = TRUE,
  longtable = FALSE,
  linesep = "",
  align = "c"
) %>%
  kableExtra::kable_styling(
    latex_options = c("striped", "repeat_header"),
    full_width = FALSE,
    position = "left",
    font_size = 9
  ) %>%
  column_spec(1, width = "3cm") %>%
  column_spec(2, width = "3cm")
if (nrow(plasmidfinder) > 0) {
    print(plasmid_tbl)
} else {
    cat("\\vspace{0.5em}")
    cat("\\textbf{Ingen treff}\n")
    cat("\\vspace{0.5em}")
}
```

\needspace{15\baselineskip}
\subsubsection*{\color{blue}5. Bioinformatisk programvare:}
\vspace{1em}

```{r results='asis', echo=FALSE}
sw_tbl <- knitr::kable(
  sw_df,
  format = "latex",
  booktabs = TRUE,
  longtable = FALSE,
  linesep = "",
  align = "l"
) %>%
  kableExtra::kable_styling(
    latex_options = c("striped", "repeat_header"),
    full_width = FALSE,
    position = "left",
    font_size = 9
  ) %>%
  column_spec(1, width = "4cm") %>%
  column_spec(2, width = "4cm")
cat("\\textbf{Software:}")
print(sw_tbl)
cat("\\vspace{1em}")

db_tbl <- knitr::kable(
  db_df,
  format = "latex",
  booktabs = TRUE,
  longtable = FALSE,
  linesep = "",
  align = "l"
) %>%
  kableExtra::kable_styling(
    latex_options = c("striped", "repeat_header"),
    full_width = FALSE,
    position = "left",
    font_size = 9
  ) %>%
  column_spec(1, width = "4cm") %>%
  column_spec(2, width = "4cm")
cat("\\textbf{Database:}")
print(db_tbl)
```



