---
output: 
  pdf_document:
    latex_engine: pdflatex
    keep_tex: true
    number_sections: false
params:
  sample_id: ""
  run_mode: ""
  genome_size: ""
  kleborate_columns: ""
  quast_out: ""
  bbmap_out: ""
  mlst_out: ""
  rmlst_out: ""
  kleborate_out: ""
  amrfinder_out: ""
  plasmidfinder_out: ""
  lrefinder_out: ""
  versions: ""
header-includes:
  - \usepackage{booktabs}
  - \usepackage{float}
  - \usepackage{graphicx}
  - \usepackage{enumitem}
  - \usepackage{xcolor}
  - \usepackage{needspace}
  - \usepackage{mathptmx}
  - \setlist[enumerate]{nosep, leftmargin=*, itemsep=0pt, topsep=0pt, partopsep=0pt, parsep=0pt}
geometry: top=1cm, bottom=2cm, left=2cm, right=2cm
---
\noindent\textcolor{violet}{\rule{\textwidth}{0.7pt}}
\subsection*{\centering \color{violet}Bioinformatic report Whole Genome Sequencing and AMR analysis}
\noindent\textcolor{violet}{\rule{\textwidth}{0.7pt}}

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(verbose = TRUE)

library(kableExtra)
library(tidyr)
library(dplyr)
library(yaml)

sample_id <- params$sample_id
```


```{r load-input-files, include=FALSE}
# Load genome size table
genome_size_tbl <- read.csv(params$genome_size, stringsAsFactors = FALSE)
kleborate_patterns <- read.csv(params$kleborate_columns, stringsAsFactors = FALSE)

# QUAST assembly stats
quast <- read.table(params$quast_out, sep = "\t", fill = TRUE, header = TRUE, comment.char = "")

# BBMap coverage
bbmap <- readLines(file.path(params$bbmap_out))


# MLST results (original mlst species name)
mlst <- read.table(params$mlst_out, sep = "\t")

# MLST (species name for report)
#mlst_newname_path=file.path("MLST", "mlst_renamed.out")
#if (file.exists(mlst_newname_path)) {
#    mlst_newname <- read.table(mlst_newname_path, sep = "\t")
#    mlst_newname_species <- mlst_newname[,2]
#}

# Ribosomal MLST
rmlst <- read.table(params$rmlst_out, sep = ":", fill = TRUE)

# Kleborate output (if present)
kleborate_files <- list.files(params$kleborate_out, full.names = TRUE)

if (length(kleborate_files) > 0) {
  kleborate <- read.table(kleborate_files, sep = "\t", header = TRUE, stringsAsFactors = FALSE, quote = "", comment.char = "")
}

# AMRFinder results
#AMR_filepath <- list.files(path = "AMRFinder", pattern = ".tsv", full.names = TRUE)
AMR_out <- read.table(params$amrfinder_out, quote = "", comment.char = "", stringsAsFactors = FALSE, header = TRUE, sep = "\t", fill = TRUE)

# Kleborate patterns (if used)
if (file.exists(params$kleborate_columns)) {
  kleborate_patterns <- read.csv(params$kleborate_columns, header = TRUE)
}

# LRE-Finder output (for Enterococcus)
if (file.exists(file.path(params$lrefinder_out))) {
  lrefinder <- readLines(file.path(params$lrefinder_out))
}


# PlasmidFinder results
plasmidfinder <- read.table(file.path(params$plasmidfinder_out), sep = "\t", header = TRUE)

# Software versions
yml <- yaml::read_yaml(params$versions)  # replace with your filename

# Optionally skip the Workflow section
yml <- yml[names(yml) != "Workflow"]

# Initialize tibbles
sw_versions <- tibble(software = character(), version = character())
db_versions <- tibble(database = character(), version = character())

for (section in names(yml)) {
  for (sw in names(yml[[section]])) {
    if (grepl("database", sw, ignore.case = TRUE)) {
      db_version <- add_row(db_versions, database = sw, version = as.character(yml[[section]][[sw]]))
    } else {
      sw_version <- add_row(sw_versions, software = sw, version = as.character(yml[[section]][[sw]]))
    }
  }
}
```

\vspace{1em}
`r format(Sys.Date(), "%d.%m.%Y")`

\vspace{1em}
\textbf{\color{violet}Sample identifier : `r sample_id`} 


\subsection*{\color{violet}Results:}

\subsubsection*{\color{blue}1. Assembly statistics:}
\vspace{1em}
```{r assembly-stats, echo=FALSE, results='asis', warning=FALSE}
quast <- quast %>%
  rename(c("Feature" = "Assembly", "Statistics" = "contigs")) %>%
  filter(Feature == "# contigs (>= 0 bp)" | Feature == "N50" | Feature == "Total length") %>%
  mutate(Feature = gsub("# contigs \\(>= 0 bp\\)", "Number of contigs", Feature))

bbmap_cov <- which(grepl("Average coverage:", bbmap))
bbmap_cov <- data.frame(
  Feature = "Coverage",
  Statistics = gsub("Average coverage: ", "", bbmap[bbmap_cov])
)


assembly_stats <- rbind(quast, bbmap_cov) %>%
  mutate(Statistics = if_else(Feature %in% c("Total length", "Number of contigs", "N50"),
                         as.character(as.integer(Statistics)),
                         as.character(Statistics)))

check_genome_size <- function(species, assembly_size, genome_size_tbl) {
  row <- genome_size_tbl[genome_size_tbl$species == species, ]
  if (nrow(row) == 0) {
    return(NA) # Species not found
  }
  lower <- row$min
  upper <- row$max
  return(assembly_size >= lower & assembly_size <= upper)
}

species <- mlst[,2]
assembly_size <- as.numeric(assembly_stats[assembly_stats$Feature == "Total length", "Statistics"])
within_limits <- check_genome_size(species, assembly_size, genome_size_tbl)

if (is.na(within_limits)) {
  genome_size_comment <- "(Lengde ikke definert i den lokale databasen)"
} else if (within_limits) {
  genome_size_comment <- "(Innenfor  akseptert genomlengde)"
} else {
  row <- genome_size_tbl[genome_size_tbl$species == species, ]
  genome_size_comment <- "(utenfor akseptert genomlengde)"
}

cat("\\begin{tabular}{@{}ll}\n")
cat("Number of contigs: & ", assembly_stats[1, 2], " \\\\\n")
cat("Assembly-length:   & ", assembly_stats[2, 2], " bp ", genome_size_comment, "\\\\\n")
cat("N50 (based on contigs \\(\\geq\\) 500 bp):               & ", assembly_stats[3, 2], " bp\\\\\n")
cat("Coverage:          & ", assembly_stats[4, 2], " \\\\\n")
cat("\\end{tabular}\n")
```

\subsubsection*{\color{blue}2. Taxonomic classification:}
\vspace{1em}

a) Species and MLST type (based on https://pubmlst.org/):
```{r mlst, echo=FALSE, results='asis', warning=FALSE}
mlst_species <- mlst[,2]
mlst_st <- mlst[,3]
```
**`r mlst_species` ST-`r mlst_st`**
\vspace{1em}

b) Ribosomal MLST - Species and support percentage (based on https://pubmlst.org/species-id):
```{r rmlst, echo=FALSE, results='asis', warning=FALSE}
rmlst <- rmlst %>%
  filter(V1 == "Taxon" | V1 == "Support")
```
**`r rmlst[1,2]`**  **`r paste0("(", rmlst[2,2], ")")`**
\vspace{1em}

```{r results='asis', echo=FALSE}
if (grepl("Klebsiella", mlst_species) && length(kleborate_files) > 0) {
  cat("c) Kleborate (based on https://github.com/katholt/Kleborate):")
  cat("\\vspace{0.5em}\n")
  kleborate_species <- kleborate[1, "enterobacterales__species__species"]
  kleborate_species_match <- kleborate[1, "enterobacterales__species__species_match"]
  kleborate_QC_warnings <- kleborate[1, "general__contig_stats__QC_warnings"]
  
  cat("\\begin{tabular}{@{}ll}\n")
  cat("Species:                 & \\textbf{", kleborate_species, "} \\\\\n")
  cat("Species match:           & \\textbf{", kleborate_species_match, "} \\\\\n")
  cat("QC warnings:             & \\textbf{", kleborate_QC_warnings, "} \\\\\n")
  cat("\\end{tabular}\n")
}
```
\newpage

\subsubsection*{\color{blue}3. Resistance:}
\vspace{1em}
Resistance genes based on the core database of the Bacterial Antimicrobial Resistance Reference Gene Database BioProject PRJNA313047. Also includes genes that confer resistance to non-antibiotics (e.g. quaternary ammonium compounds) and antibiotics used in veterinary medicine.

```{r, include=FALSE}
AMR_out <- AMR_out[, c("Scope", "Type", "Subtype", "Element.symbol", "Class", "Subclass", "Method", "X..Coverage.of.reference", "X..Identity.to.reference")]
colnames(AMR_out) <- c("Scope", "Type", "Subtype", "Element", "Class", "Subclass", "Method", "Coverage", "Identity")
AMR_out$Class <- gsub('/',' ',AMR_out$Class)
AMR_out$Subclass <- gsub("/", " ", AMR_out$Subclass)

AMR_1st_criteria <- AMR_out %>%
  filter(Scope == "core" | Scope == "plus" & Type == "AMR") %>%
  filter(Coverage >= 90 & Identity >= 90) %>%
  arrange(Class, Subclass) %>%
  select(Element, Class, Subclass, Coverage, Identity)
```

\vspace{1em}
\textbf{a) Criteria:} $\,\geq\,$ 90\% coverage and $\,\geq\,$ 90\% identity

```{r AMRFinder_test, echo=FALSE, results='asis'}
knitr::kable(
    AMR_1st_criteria,
    format = "latex",
    booktabs = TRUE,
    linesep = "",
    align = "l",
    ) %>% 
  kableExtra::kable_styling(
    latex_options = c("striped", "repeat_header"),
    full_width = FALSE,
    position = "left",
    font_size = 9,
    ) %>%
  column_spec(1, width = "2cm") %>%
  column_spec(2, width = "4cm") %>%
  column_spec(3, width = "5cm") %>%
  column_spec(4, width = "1cm") %>%
  column_spec(5, width = "1cm")
```
\vspace{1em}
\needspace{15\baselineskip}
\textbf{b) Mutations-mediated resistance:}

```{r, echo=FALSE, results='asis'}
AMR_2nd_criteria <- AMR_out %>%
  filter(Method == "POINTX") %>%
  arrange(Class, Subclass) %>%
  select(Element, Class, Subclass, Coverage, Identity)

knitr::kable(
  AMR_2nd_criteria,
    format = "latex",
    booktabs = TRUE,
    linesep = "",
    align = "l",
    ) %>% 
  kableExtra::kable_styling(
    latex_options = c("striped", "repeat_header"),
    full_width = FALSE,
    position = "left",
    font_size = 9,
    ) %>%
  column_spec(1, width = "2cm") %>%
  column_spec(2, width = "4cm") %>%
  column_spec(3, width = "5cm") %>%
  column_spec(4, width = "1cm") %>%
  column_spec(5, width = "1cm")
```
\vspace{1em}
\needspace{15\baselineskip}
\textbf{c) Criteria:} $\,\geq\,$ 60\% - $\,\leq\,$ 90\% coverage and $\,\geq\,$ 60\% - $\,\leq\,$ 90\% nucleotide identity

```{r, echo=FALSE, results='asis'}
AMR_3rd_criteria <- AMR_out %>%
  filter(Scope == "core" | Scope == "plus" & Type == "AMR") %>%
  filter(Coverage < 90 | Identity < 90) %>%
  filter(Coverage >= 90 & Identity >= 60 | Coverage >= 60 & Identity >= 90) %>%
  arrange(Class, Subclass) %>%
  select(Element, Class, Subclass, Coverage, Identity)

knitr::kable(
  AMR_3rd_criteria,
    format = "latex",
    booktabs = TRUE,
    linesep = "",
    align = "l",
    ) %>% 
  kableExtra::kable_styling(
    latex_options = c("striped", "repeat_header"),
    full_width = FALSE,
    position = "left",
    font_size = 9,
    ) %>%
  column_spec(1, width = "2cm") %>%
  column_spec(2, width = "4cm") %>%
  column_spec(3, width = "5cm") %>%
  column_spec(4, width = "1cm") %>%
  column_spec(5, width = "1cm")
```
\vspace{1em}
\needspace{15\baselineskip}
\textbf{d) Genes associated with resistance, biocides, stress, efflux, virulence:} $\,\geq\,$ 60\% coverage and $\,\geq\,$ 60\% nucleotide identity.
Genes based on the Plus database of the Bacterial Antimicrobial Resistance Reference Gene Database BioProject PRJNA313047.

```{r, echo=FALSE, results='asis'}
AMR_4th_criteria <- AMR_out %>%
  filter(Scope == "plus" & !Type == "AMR") %>%
  arrange(Class, Subclass) %>%
  select(Element, Type, Subtype, Class, Subclass, Coverage, Identity)
  
knitr::kable(
  AMR_4th_criteria,
    format = "latex",
    booktabs = TRUE,
    longtable = TRUE,
    linesep = "",
    align = "l",
    ) %>% 
  kableExtra::kable_styling(
    latex_options = c("striped", "repeat_header"),
    full_width = FALSE,
    position = "left",
    font_size = 9,
    ) %>%
  column_spec(1, width = "2cm") %>%
  column_spec(2, width = "2.3cm") %>%
  column_spec(3, width = "2.3cm") %>%
  column_spec(4, width = "2.3cm") %>%
  column_spec(5, width = "2.3cm") %>%
  column_spec(6, width = "1cm") %>%
  column_spec(7, width = "1cm")
```

\vspace{1em}
```{r results='asis', echo=FALSE}
if (grepl("Klebsiella", mlst_species) && length(kleborate_files) > 0) {
  cat("\\needspace{15\\baselineskip}\n")
  cat("\\textbf{e) Klebsiella-specific resistance analysis:}\n")
  # Prepare the Kleborate data
  kleborate_long <- tibble(
    Pattern = names(kleborate),
    Value = as.character(unlist(kleborate[1, ]))
  )
  
  kleborate_long <- kleborate_long %>%
    mutate(Pattern = gsub(".*__", "", Pattern))
  
  kleborate_long_filtered <- kleborate_long[kleborate_long$Pattern %in% kleborate_patterns$Pattern, ]
  kleborate_long_filtered_metadata <- kleborate_long_filtered %>%
    left_join(kleborate_patterns, by = "Pattern") %>%
    arrange(Type)
  
  # Separate tables by Type
  mutations_table <- kleborate_long_filtered_metadata %>%
    filter(Type == "mutations") %>%
    select(Pattern, Value)
  
  virulence_table <- kleborate_long_filtered_metadata %>%
    filter(Type == "virulence") %>%
    select(Pattern, Value)
  
  locus_table <- kleborate_long_filtered_metadata %>%
    filter(Type %in% c("K_locus", "O_locus")) %>%
    select(Pattern, Value)
  
  # Render the mutations table
  cat("\\vspace{1em}\n\\par\n")
  cat("\\needspace{15\\baselineskip}\n")
  cat("\\textbf{Mutasjoner:}\n")
  knitr::kable(
    mutations_table,
    format = "latex",
    booktabs = TRUE,
    longtable = FALSE,
    linesep = "",
    align = "l"
  ) %>%
    kableExtra::kable_styling(
      latex_options = c("striped", "repeat_header"),
      full_width = FALSE,
      position = "left",
      font_size = 9
    ) %>%
    column_spec(1, width = "5cm") %>%
    column_spec(2, width = "4cm") %>%
    print()
    cat("Colistin resistance is determined by truncation of/or absence of mgrB and pmrB core genes. OmpK35 and OmpK36 resistance is determined by truncation or point mutations. Truncation is represented by % length of amino acids from codon start.\n\\par\n")

  # Render the virulence table
  cat("\\vspace{1em}\n")
  cat("\\needspace{15\\baselineskip}\n")
  cat("\\textbf{Virulence factors:}\n")
  knitr::kable(
    virulence_table,
    format = "latex",
    booktabs = TRUE,
    longtable = FALSE,
    linesep = "",
    align = "l"
  ) %>%
    kableExtra::kable_styling(
      latex_options = c("striped", "repeat_header"),
      full_width = FALSE,
      position = "left",
      font_size = 9
    ) %>%
    column_spec(1, width = "5cm") %>%
    column_spec(2, width = "4cm") %>%
    print()
    cat("0 = negative for all of yersiniabactin (ybt), colibactin (clb), aerobactin (iuc); 1 = yersiniabactin only; 2 = yersiniabactin and colibactin (or colibactin only); 3 = aerobactin (without yersiniabactin or colibactin); 4 = aerobactin with yersiniabactin (without colibactin); 5 = yersiniabactin, colibactin and aerobactin \n\\par\n")
  
  # Render the O and K locus table
  cat("\\vspace{1em}\n")
  cat("\\needspace{15\\baselineskip}\n")
  cat("\\textbf{O- og K-locus:}\n")
  knitr::kable(
    locus_table,
    format = "latex",
    booktabs = TRUE,
    longtable = FALSE,
    linesep = "",
    align = "l"
  ) %>%
    kableExtra::kable_styling(
      latex_options = c("striped", "repeat_header"),
      full_width = FALSE,
      position = "left",
      font_size = 9
    ) %>%
    column_spec(1, width = "5cm") %>%
    column_spec(2, width = "4cm") %>%
    print()
}
```

\vspace{1em}
```{r results='asis', echo=FALSE}
if (grepl("Enterococcus", mlst_species)) {
  cat("\\needspace{20\\baselineskip}\n")
  cat("\\textbf{e) Linezolid Resistance: (based on LRE-finder, https://cge.cbs.dtu.dk/services/LRE-finder/)}")
  cat("\\vspace{1em}\n")

  # Species identified
  lrefinder_species <- which(grepl("Species identified:", lrefinder)) + 1
  lrefinder_species <- lrefinder[lrefinder_species]

  cat("\\begin{tabular}{@{}ll}\n")
  cat("Species:                 & \\textbf{", lrefinder_species, "} \\\\\n")
  cat("\\end{tabular}\n")

  # Find section starts
  genes_start <- which(grepl("Genes Identified:", lrefinder)) + 1
  mut_start <- which(grepl("Identified mutations in 23s:", lrefinder)) + 1

  cat("\\vspace{1em}\n")
  cat("Gener:\n")

  # Extract gene hits
  gene_lines <- lrefinder[genes_start:(mut_start - 3)]
  gene_lines <- gene_lines[gene_lines != ""]
  if (length(gene_lines) > 1) {  # header + at least one data line
    gene_df <- read.table(text = paste(gene_lines, collapse = "\n"), header = TRUE, stringsAsFactors = FALSE)
    lre_gene <- knitr::kable(
      gene_df,
      format = "latex",
      booktabs = TRUE,
      longtable = FALSE,
      linesep = "",
      align = "c"
    ) %>%
      kableExtra::kable_styling(
        latex_options = c("striped", "repeat_header"),
        full_width = FALSE,
        position = "left",
        font_size = 9
      ) %>%
      column_spec(1, width = "4cm") %>%
      column_spec(2, width = "2cm") %>%
      column_spec(3, width = "2cm")
    print(lre_gene)

  } else {
    cat("Ingen treff\n")
  }

  # Extract mutations
  cat("Mutations:\n")
  mut_lines <- lrefinder[(mut_start + 1):length(lrefinder)]
  mut_lines <- mut_lines[mut_lines != "" & !grepl("reference", mut_lines)]
  if (length(mut_lines) > 0) {
    mut_df <- read.table(text = paste(mut_lines, collapse = "\n"), header = FALSE, stringsAsFactors = FALSE)
    colnames(mut_df) <- c("Position", "WildTypeRatio", "MutantRatio", "Phenotype")
    lre_mut <- knitr::kable(
      mut_df,
      format = "latex",
      booktabs = TRUE,
      longtable = FALSE,
      linesep = "",
      align = "c"
    ) %>%
      kableExtra::kable_styling(
        latex_options = c("striped", "repeat_header"),
        full_width = FALSE,
        position = "left",
        font_size = 9
      ) %>%
      column_spec(1, width = "2cm") %>%
      column_spec(2, width = "2cm") %>%
      column_spec(3, width = "1.5cm") %>%
      column_spec(4, width = "1.5cm")
    print(lre_mut)
  } else {
    print("Ingen treff")
  }
}
```

\needspace{10\baselineskip}
\subsubsection*{\color{blue}4. Plasmid replicon:}
\vspace{1em}
(based on PlasmidFinder https://cge.cbs.dtu.dk/services/PlasmidFinder/). Analysis only applies to Enterobacterales and Gram-positive.

```{r results='asis', echo=FALSE}
plasmidfinder <- plasmidfinder %>%
  select(Plasmid, Identity)

plasmid_tbl <- knitr::kable(
  plasmidfinder,
  format = "latex",
  booktabs = TRUE,
  longtable = FALSE,
  linesep = "",
  align = "c"
) %>%
  kableExtra::kable_styling(
    latex_options = c("striped", "repeat_header"),
    full_width = FALSE,
    position = "left",
    font_size = 9
  ) %>%
  column_spec(1, width = "3cm") %>%
  column_spec(2, width = "3cm")
print(plasmid_tbl)
```

\needspace{15\baselineskip}
\subsubsection*{\color{blue}5. Bioinformatics software:}
\vspace{1em}

```{r results='asis', echo=FALSE}
sw_tbl <- knitr::kable(
  sw_versions,
  format = "latex",
  booktabs = TRUE,
  longtable = FALSE,
  linesep = "",
  align = "c"
) %>%
  kableExtra::kable_styling(
    latex_options = c("striped", "repeat_header"),
    full_width = FALSE,
    position = "left",
    font_size = 9
  ) %>%
  column_spec(1, width = "3cm") %>%
  column_spec(2, width = "3cm")
print(sw_tbl)

db_tbl <- knitr::kable(
  db_version,
  format = "latex",
  booktabs = TRUE,
  longtable = FALSE,
  linesep = "",
  align = "c"
) %>%
  kableExtra::kable_styling(
    latex_options = c("striped", "repeat_header"),
    full_width = FALSE,
    position = "left",
    font_size = 9
  ) %>%
  column_spec(1, width = "3cm") %>%
  column_spec(2, width = "3cm")
print(db_tbl)
```



